Add new `create_inits()` methods to other stanfit classes

amas0 opened the pull request:
#### Submission Checklist

- [ x ] Run unit tests
- [ x ] Declare copyright holder and open-source license: see below

#### Summary

This PR addresses #745. Implementation largely mirrors that for `CmdStanPathfinder`. In particular, this adds the following methods: `CmdStanVB.create_inits()`, `CmdStanLaplace.create_inits()`, `CmdStanMLE.create_intis()`, `CmdStanMCMC.create_inits()`. With the exception of MLE, inits are selected by sampling without replacement from draws. The MLE implementation will initialize all requested chains at the optimized parameter values.

Unit tests for all these methods are added as well.

I also updated the User's Guide example on creating inits from VI outputs to be more general and cover these new methods.

I assume this is already well-known but it seems like there is a cleaner library structure to have each of these different stanfit objects be implementations of the same parent (or abstract) class to unify these implementations (and those of other common methods). I haven't worked through the details, but it seemed possible given how similar the internal structures tend to be.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): myself



By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

amas0 committed:
Add CmdStanMCMC.create_inits()

amas0 committed:
Fix inconsistent string quoting

amas0 committed:
Add CmdStanMCMC.create_inits() test

amas0 committed:
Add CmdStanLaplace.create_inits()

amas0 committed:
Add CmdStanMLE.create_inits()

amas0 committed:
Add CmdStanVB.create_inits()

amas0 committed:
Add sampling tests from *.create_inits() for MLE, VB, Laplace, and MCMC

amas0 committed:
Update sampler init user's guide example

WardBrian commented on cmdstanpy/stanfit/laplace.py:
```diff
@@ -52,6 +52,39 @@ def __init__(self, runset: RunSet, mode: CmdStanMLE) -> None:
+
+        If ``chains`` is 1, a dictionary is returned, otherwise a list
+        of dictionaries is returned, in the format expected for the
+        ``inits`` argument. of :meth:`CmdStanModel.sample`.
```
I think this typo is also in Pathfinder's doc, but the period after "argument" should be removed

amas0 replied:
The perils of copy/paste -- good catch.

WardBrian commented on cmdstanpy/stanfit/mle.py:
```diff
@@ -36,6 +36,30 @@ def __init__(self, runset: RunSet) -> None:
+        if chains == 1:
+            return mle_inits
+        else:
+            return [mle_inits for _ in range(chains)]
```
Because the default behavior is if you pass a dictionary to sample, it uses it for every chain, I don't know if this is super necessary. 

I do think this function should probably accept both a seed and chains argument, even if they're both unused, to let people write more generic code

WardBrian replied:
(I also think this function might be exactly equivalent to `self.stan_variables()` after those changes?

amas0 replied:
Updated the signature for compatibility. I also use `stan_variables()` in the function now, but because `stan_variables()` will currently return a `float` if the parameter is scalar, I just map everything to an `np.array` for type compatibility. It seems this is intended to be the future behavior of `stan_variables()` in 2.0, at which point it will be equivalent.

WardBrian commented on docsrc/users-guide/examples/VI as Sampler Inits.ipynb:
I think we don't generally check in the output of the notebooks

amas0 replied:
Good catch, fixed. It might be worth adding a pre-commit hook for this using `nbstripout`?

WardBrian replied:
I think there is one that takes too long to run so we *do* check it in, otherwise we probably could do with a pre-commit

WardBrian reviewed (commented):
Thanks! Overall this looks great, I just have a few minor comments.

Re:
> I assume this is already well-known but it seems like there is a cleaner library structure to have each of these different stanfit objects be implementations of the same parent (or abstract) class to unify these implementations (and those of other common methods).

Yes, we definitely could. So far most of what we've done in this style is using helper methods we throw in the utils subpackage for code re-use, but inheritance would also get us there, and maybe a bit more cleanly. In another project that deals with Stan outputs, I don't even differentiate between the different algorithms, I just have one 'StanOutput' class that can hold any of them.

amas0 committed:
Fix errant period in docstrings

amas0 committed:
Update CmdStanMLE.create_inits() compatibility

This change simplifies the create_inits() for the MLE
params to always return a dictionary of inits, rather
than a list for multiple chains. The default behavior
of the sample method is to initialize all chains
at the same init if only one is given per param.

The chains paramter is kept and the seed parameter is
added to the signature, despite being no-ops, for
the purposes of maintaining uniformity across the
other create_inits() methods on other stanfit objects.

amas0 committed:
Strip output from sampler inits notebook

amas0 commented:
Thanks for the review -- responded to your comments and made changes where appropriate.

Not sure what's going on with that failed pytest check. Some of the cancelled jobs seem to have gotten through the testing stage without issue.

Going through the library, there are a number of references to a 2.0 version -- that might be a good opportunity to clean up some of these stanfit objects and such. Not sure if there are explicit plans for what that will contain (or timeline), just a thought. Happy to help out if there's a direction there.

WardBrian commented:
Because ADVI can be unstable it might be worth setting a known-good seed in the new test that runs it

amas0 committed:
Fix test_variational test on creating inits

amas0 commented:
Okay, pinned some seeds that seemed fine and checks look good. Serendipitous that I went back and did that, because I found a bug in one of the tests I wrote, which I now have fixed.

I think this should be set?

bob-carpenter commented:
Thanks.  @mitzimorris may be able to review and approve sooner than @WardBrian for this one.

mitzimorris commented:
the code looks great.

the documentation section on creating inits should also mention creating inits from previous MCMC runs.
I would be happy merging this now and doing a follow-on PR to improve the docs.

mitzimorris commented on docsrc/users-guide/examples/VI as Sampler Inits.ipynb:
```diff
@@ -4,13 +4,16 @@
    "metadata": {},
    "source": [
-    "## Using Variational Estimates to Initialize the NUTS-HMC Sampler\n",
+    "## Using Estimates from Variational, Laplace, or Optimization Methods to Initialize the NUTS-HMC Sampler\n",
```
we should rename this page altogether - "Initializing the NUTS-HMC sampler"

amas0 replied:
Updated.

mitzimorris commented on docsrc/users-guide/examples/VI as Sampler Inits.ipynb:
```diff
@@ -4,13 +4,16 @@
-    "[ADVI ](https://mc-stan.org/docs/cmdstan-guide/variational_config.html) \n",
-    "as the initial parameter values for Stan's NUTS-HMC sampler.\n",
-    "By default, the sampler algorithm randomly initializes all model parameters in the range uniform\\[-2, 2\\].  When the true parameter value is outside of this range, starting from the estimates from Pathfinder and ADVI will speed up and improve adaptation.\n",
+    "In this example, we show how to use parameter estimates returned by Stan's various posterior approximation or optimization algorithms as initial values for Stan's NUTS-HMC sampler. These include:\n",
```
In this example we show how to use the returned estimates from a previous run of any of the inference algorithms can be used as initial values for  Stan's NUTS-HMC sampler.

amas0 replied:
Updated.

mitzimorris reviewed (approved):
This could be done as a follow-on PR - the code looks good.

amas0 commented:
I'm happy to quickly update the docs as part of this PR. I do mention that previous MCMC estimates can be used as inits, but it's more of an aside at the end. I'll make it clearer that it's included.

amas0 committed:
Clarify docs on using sampler output as inits

amas0 commented:
Updated the docs to clarify that one can use the output of one MCMC run as inits into another.

One relatively minor point. With this update, the file name of the example notebook `VI as Sampler Inits.ipynb` is somewhat outdated. I think it would make sense to update this, but doing so would break any existing links to that example as the URL is generated from the file name. I imagine a redirect could be set up with Sphinx somehow.

mitzimorris commented:
you're right - we're stuck with the filename.

mitzimorris reviewed (approved):
Thanks - this all looks great!
